frontend/src/app/main/data/workspace/selection.cljs

```diff
@@ -332,54 +333,81 @@
-        update-unames! (fn [new-name] (vswap! unames conj new-name))
-        all-ids        (reduce #(into %1 (cons %2 (cph/get-children-ids all-objects %2))) (d/ordered-set) ids)
-        ids-map        (into {} (map #(vector % (uuid/next))) all-ids)
-
-        init-changes
-        (-> (pcb/empty-changes it)
-            (pcb/with-page page)
-            (pcb/with-objects all-objects))
-
-        changes
-        (->> shapes
-             (reduce #(prepare-duplicate-shape-change %1
-                                                      all-objects
-                                                      page
-                                                      unames
-                                                      update-unames!
-                                                      ids-map
-                                                      %2
-                                                      delta
-                                                      libraries)
-                     init-changes))]
-
-    (-> changes
-        (prepare-duplicate-flows shapes page ids-map)
-        (prepare-duplicate-guides shapes page ids-map delta))))
+  ([all-objects page ids delta it libraries state]
+   (let [init-changes
+         (-> (pcb/empty-changes it)
+             (pcb/with-page page)
+             (pcb/with-objects all-objects))]
+  (prepare-duplicate-changes all-objects page ids delta it libraries state init-changes)))
+
+  ([all-objects page ids delta it libraries state init-changes]
+   (let [shapes         (map (d/getf all-objects) ids)
+         unames         (volatile! (cp/retrieve-used-names (:objects page)))
+         update-unames! (fn [new-name] (vswap! unames conj new-name))
+         all-ids        (reduce #(into %1 (cons %2 (cph/get-children-ids all-objects %2))) (d/ordered-set) ids)
+         ids-map        (into {} (map #(vector % (uuid/next))) all-ids)
+
+
+
+         changes
+         (->> shapes
+              (reduce #(prepare-duplicate-shape-change %1
+                                                       all-objects
+                                                       page
+                                                       unames
+                                                       update-unames!
+                                                       ids-map
+                                                       %2
+                                                       delta
+                                                       libraries
+                                                       state
+                                                       it)
+                      init-changes))]
+
+     (-> changes
+         (prepare-duplicate-flows shapes page ids-map)
+         (prepare-duplicate-guides shapes page ids-map delta)))))
+
+(defn- prepare-duplicate-component-change
+  ([changes page component delta libraries state it]
```

Comment:
The attribute `component` should be renamed to `component-root`, because it's a shape, not a component object.